File: languages/elixir/k-means-par.exs
# k-means-par: parallel scaling-track variant.
# Invocation: elixir k-means-par.exs <cores> <n>
# Decomposition: ITERS iterations; per iteration:
#   1. Parallel ASSIGNMENT: divide N points into `cores` contiguous bands.
#      Each worker computes assignments for its points (strict < lowest-index tie-break
#      preserved -- workers process points in order within their band).
#      Each worker returns:
#        - a list of {point_index, cluster_index} in ascending order (for writing assign)
#        - partial ssum list (K*D) and cnt list (K) accumulated over its band
#   2. Serial CENTROID UPDATE: merge partial sums from all workers; floor-mean;
#      empty-cluster unchanged (identical to serial).
#   After all tasks complete, write assignments into the shared :atomics array
#   (each worker wrote only its disjoint index range).
# Core-invariant: each point's assignment depends only on point coords + current centroids
# (both read-only during assignment). Band boundaries via floor-division are deterministic.
# Output is identical to the serial benchmark for cores=1,2,4.
import Bitwise

defmodule KMeansPar do
  @p 1_000_000_007
  @k 16
  @d 4
  @iters 10
  @range 256

  defp lcg(s), do: band(s * 1_103_515_245 + 12_345, 0x7FFFFFFF)

  defp aget(ref, i), do: :atomics.get(ref, i + 1)
  defp aput(ref, i, v), do: :atomics.put(ref, i + 1, v)

  defp fill(_pt, i, total, _state) when i >= total, do: :ok

  defp fill(pt, i, total, state) do
    state = lcg(state)
    aput(pt, i, rem(state, @range))
    fill(pt, i + 1, total, state)
  end

  defp init_cen(_pt, _cen, i) when i >= @k * @d, do: :ok

  defp init_cen(pt, cen, i) do
    aput(cen, i, aget(pt, i))
    init_cen(pt, cen, i + 1)
  end

  defp dist2(_pt, _cen, _pb, _cb, d, acc) when d >= @d, do: acc

  defp dist2(pt, cen, pb, cb, d, acc) do
    df = aget(pt, pb + d) - aget(cen, cb + d)
    dist2(pt, cen, pb, cb, d + 1, acc + df * df)
  end

  defp nearest(_pt, _cen, _i, k, best, _bd) when k >= @k, do: best

  defp nearest(pt, cen, i, k, best, bd) do
    dist = dist2(pt, cen, i * @d, k * @d, 0, 0)

    if bd < 0 or dist < bd do
      nearest(pt, cen, i, k + 1, k, dist)
    else
      nearest(pt, cen, i, k + 1, best, bd)
    end
  end

  # Compute dimensions sum contribution for point at pb into ssum at cluster offset sb.
  defp add_dims_list(_pt, ssum, _pb, _sb, d) when d >= @d, do: ssum

  defp add_dims_list(pt, ssum, pb, sb, d) do
    val = aget(pt, pb + d)
    ssum2 = List.update_at(ssum, sb + d, &(&1 + val))
    add_dims_list(pt, ssum2, pb, sb, d + 1)
  end

  # Process points [i_start, i_end): assign each point and accumulate partial sums.
  # Returns: {assign_list (list of cluster indices, ascending i), ssum_list (K*D), cnt_list (K)}.
  defp process_band(_pt, _cen, i, i_end, assign_acc, ssum, cnt) when i >= i_end,
    do: {Enum.reverse(assign_acc), ssum, cnt}

  defp process_band(pt, cen, i, i_end, assign_acc, ssum, cnt) do
    c = nearest(pt, cen, i, 0, 0, -1)
    cnt2 = List.update_at(cnt, c, &(&1 + 1))
    ssum2 = add_dims_list(pt, ssum, i * @d, c * @d, 0)
    process_band(pt, cen, i + 1, i_end, [c | assign_acc], ssum2, cnt2)
  end

  # Write a list of cluster indices into assign atomics starting at i_start.
  defp write_assignments(_assign, _i, []), do: :ok

  defp write_assignments(assign, i, [c | rest]) do
    aput(assign, i, c)
    write_assignments(assign, i + 1, rest)
  end

  # Merge partial sums/counts from all workers.
  defp merge_partials(partial_results) do
    ssum_init = List.duplicate(0, @k * @d)
    cnt_init = List.duplicate(0, @k)

    Enum.reduce(partial_results, {ssum_init, cnt_init}, fn {_assigns, pssum, pcnt},
                                                            {ssum, cnt} ->
      ssum2 = Enum.zip_with(ssum, pssum, fn a, b -> a + b end)
      cnt2 = Enum.zip_with(cnt, pcnt, fn a, b -> a + b end)
      {ssum2, cnt2}
    end)
  end

  defp recompute(_cen, _ssum, _cnt, k) when k >= @k, do: :ok

  defp recompute(cen, ssum, cnt, k) do
    c = Enum.at(cnt, k)
    if c > 0, do: mean_dims_list(cen, ssum, c, k * @d, 0)
    recompute(cen, ssum, cnt, k + 1)
  end

  defp mean_dims_list(_cen, _ssum, _c, _kb, d) when d >= @d, do: :ok

  defp mean_dims_list(cen, ssum, c, kb, d) do
    aput(cen, kb + d, div(Enum.at(ssum, kb + d), c))
    mean_dims_list(cen, ssum, c, kb, d + 1)
  end

  defp iterate(_pt, _cen, _assign, _cores, n, iter) when iter >= @iters, do: :ok

  defp iterate(pt, cen, assign, cores, n, iter) do
    # Parallel: each worker assigns + accumulates its band.
    tasks =
      Enum.map(0..(cores - 1), fn w ->
        i_start = div(w * n, cores)
        i_end = div((w + 1) * n, cores)
        ssum_init = List.duplicate(0, @k * @d)
        cnt_init = List.duplicate(0, @k)

        Task.async(fn ->
          process_band(pt, cen, i_start, i_end, [], ssum_init, cnt_init)
        end)
      end)

    # Collect results in band order.
    band_results = Enum.map(tasks, fn t -> Task.await(t, :infinity) end)

    # Write assignments into shared atomics (each band writes its own disjoint range).
    Enum.each(Enum.zip(0..(cores - 1), band_results), fn {w, {assigns, _ssum, _cnt}} ->
      i_start = div(w * n, cores)
      write_assignments(assign, i_start, assigns)
    end)

    # Serial centroid update.
    {merged_ssum, merged_cnt} = merge_partials(band_results)
    recompute(cen, merged_ssum, merged_cnt, 0)
    iterate(pt, cen, assign, cores, n, iter + 1)
  end

  # Serial final assignment (matches serial benchmark's assign_all).
  defp assign_all_serial(_pt, _cen, _assign, i, n) when i >= n, do: :ok

  defp assign_all_serial(pt, cen, assign, i, n) do
    best = nearest(pt, cen, i, 0, 0, -1)
    aput(assign, i, best)
    assign_all_serial(pt, cen, assign, i + 1, n)
  end

  defp hash_cen(_cen, i, h) when i >= @k * @d, do: h

  defp hash_cen(cen, i, h) do
    hash_cen(cen, i + 1, rem(h * 31 + aget(cen, i), @p))
  end

  defp hash_assign(_assign, i, n, h) when i >= n, do: h

  defp hash_assign(assign, i, n, h) do
    hash_assign(assign, i + 1, n, rem(h * 31 + aget(assign, i), @p))
  end

  def run(cores, n) do
    pt = :atomics.new(max(n * @d, 1), signed: true)
    cen = :atomics.new(@k * @d, signed: true)
    assign = :atomics.new(max(n, 1), signed: true)

    fill(pt, 0, n * @d, 42)
    init_cen(pt, cen, 0)
    t0 = System.monotonic_time(:nanosecond)
    iterate(pt, cen, assign, cores, n, 0)
    t1 = System.monotonic_time(:nanosecond)
    IO.puts(:stderr, "COMPUTE_NS #{t1 - t0}")
    # Final assignment with the final centroids, then checksum (identical to serial).
    assign_all_serial(pt, cen, assign, 0, n)
    h = hash_cen(cen, 0, 0)
    hash_assign(assign, 0, n, h)
  end
end

[cores_s, n_s] =
  case System.argv() do
    [c, n | _] -> [c, n]
    [c] -> [c, "8000"]
    _ -> ["1", "8000"]
  end

cores = String.to_integer(cores_s)
n = String.to_integer(n_s)

IO.puts(KMeansPar.run(cores, n))
IO.puts("k-means(#{n})")

File: languages/go/k-means-par.go
// k-means-par: parallel Lloyd's k-means clustering.
// Invocation: k-means-par <cores> <n>
// Same algorithm and checksum as serial k-means; parallelises the ASSIGNMENT
// step of each iteration (and the final assignment pass) by partitioning the N
// points into `cores` contiguous bands. Each goroutine writes assign[i] for its
// own range of i — disjoint writes, no synchronisation needed on the array.
// The centroid UPDATE step (ssum accumulation + floor-mean) runs serially from
// the main goroutine after the barrier, identical to the serial version.
// Tie-break rule (strict <, lowest-index centroid wins) is preserved: workers
// scan k=0..K-1 in order and the partition is contiguous, so point order is
// unchanged.
package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
	"time"
)

const (
	P     = 1000000007
	K     = 16
	D     = 4
	ITERS = 10
	RANGE = 256
)

func lcg(s int64) int64 { return (s*1103515245 + 12345) & 0x7fffffff }

// assignBand computes the nearest centroid for points [start, end) and writes
// results into assign[start..end). cen is read-only and shared.
func assignBand(pt, cen []int64, assign []int, start, end int) {
	for i := start; i < end; i++ {
		best := 0
		bd := int64(-1)
		for k := 0; k < K; k++ {
			var dist int64 = 0
			for d := 0; d < D; d++ {
				df := pt[i*D+d] - cen[k*D+d]
				dist += df * df
			}
			if bd < 0 || dist < bd {
				bd = dist
				best = k
			}
		}
		assign[i] = best
	}
}

func run(cores, n int) int64 {
	pt := make([]int64, n*D)
	s := int64(42)
	for i := 0; i < n*D; i++ {
		s = lcg(s)
		pt[i] = s % RANGE
	}
	cen := make([]int64, K*D)
	for i := 0; i < K*D; i++ {
		cen[i] = pt[i] // initial centroids = first K points
	}
	assign := make([]int, n)

	var wg sync.WaitGroup

	// parallelAssign runs the assignment step over all N points.
	parallelAssign := func() {
		wg.Add(cores)
		for w := 0; w < cores; w++ {
			w := w
			start := w * n / cores
			end := (w + 1) * n / cores
			go func() {
				defer wg.Done()
				assignBand(pt, cen, assign, start, end)
			}()
		}
		wg.Wait()
	}

	t0 := time.Now()
	for iter := 0; iter < ITERS; iter++ {
		parallelAssign()

		// Serial centroid update: floor-mean, empty cluster unchanged.
		ssum := make([]int64, K*D)
		cnt := make([]int64, K)
		for i := 0; i < n; i++ {
			k := assign[i]
			cnt[k]++
			for d := 0; d < D; d++ {
				ssum[k*D+d] += pt[i*D+d]
			}
		}
		for k := 0; k < K; k++ {
			if cnt[k] > 0 {
				for d := 0; d < D; d++ {
					cen[k*D+d] = ssum[k*D+d] / cnt[k]
				}
			}
		}
	}

	// Final assignment pass with final centroids (same as serial).
	parallelAssign()
	fmt.Fprintf(os.Stderr, "COMPUTE_NS %d\n", time.Since(t0).Nanoseconds())

	var h int64 = 0
	for i := 0; i < K*D; i++ {
		h = (h*31 + cen[i]) % P
	}
	for i := 0; i < n; i++ {
		h = (h*31 + int64(assign[i])) % P
	}
	return h
}

func main() {
	cores := 1
	n := 8000
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil {
			cores = v
		}
	}
	if len(os.Args) > 2 {
		if v, err := strconv.Atoi(os.Args[2]); err == nil {
			n = v
		}
	}
	runtime.GOMAXPROCS(cores)
	fmt.Println(run(cores, n))
	fmt.Printf("k-means(%d)\n", n)
}
